### PR TITLE
cephadm: support upgrade from octopus to pacific

### PR DESCRIPTION
**Pull request by liewegas**
Implement basic version checks to support a major release upgrade.

Complete the upgrade by setting require-osd-release for the OSDs.

### CODE REVIEW

File: src/pybind/mgr/cephadm/upgrade.py [hunk 1]
```diff
@@ -90,19 +90,50 @@ def upgrade_status(self) -> orchestrator.UpgradeStatusSpec:
...
+        # to far a jump?
+        current_version = self.mgr.version.split('ceph version ')[1]
+        current_major, current_minor, current_patch = current_version.split('-')[0].split('.')
+        if int(current_major) < int(major) - 2:
+            return f'ceph can only upgrade 1 or 2 major versions at a time; {current_version} -> {version} is too big a jump'
+        if int(current_major) > int(major):
+            return f'ceph cannot downgrade major versions (from {current_version} to {version})'
+        if int(current_major) == int(major):
+            if int(current_minor) > int(minor):
+                return f'ceph cannot downgrade to a {"rc" if minor == "1" else "dev"} release'
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/upgrade.py:113]: :-1: for special handing of rc and dev releases. 

```suggestion
                return f'ceph cannot downgrade versions (from {current_version} to {version})'
```

File: src/pybind/mgr/cephadm/upgrade.py [hunk 2]
```diff
@@ -90,19 +90,50 @@ def upgrade_status(self) -> orchestrator.UpgradeStatusSpec:
...
+        # check mon min
+        monmap = self.mgr.get("mon_map")
+        mon_min = monmap.get("min_mon_release", 0)
+        if mon_min < int(major) - 2:
+            return f'min_mon_release ({mon_min}) < target {major} - 2; first complete an upgrade to an earlier release'
+
+        # check osd min
+        osdmap = self.mgr.get("osd_map")
+        osd_min_name = osdmap.get("require_osd_release", "argonaut")
+        osd_min = ord(osd_min_name[0]) - ord('a') + 1
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/upgrade.py:124]:
```pyhton
def ceph_codename_to_major(codename):
    """argonaut -> 1, bobtail ... ocotpus -> 15 etc."""
    return ord(osd_min_name[0]) - ord('a') + 1
```

```suggestion
        osd_min = ceph_codename_to_major(osd_min_name)
```

File: src/pybind/mgr/cephadm/upgrade.py [hunk 3]
```diff
@@ -90,19 +90,50 @@ def upgrade_status(self) -> orchestrator.UpgradeStatusSpec:
...
-                (major, minor, patch) = version.split('.')
-                assert int(minor) >= 0
-                assert int(patch) >= 0
-            except:
-                raise OrchestratorError('version must be in the form X.Y.Z (e.g., 15.2.3)')
-            if int(major) < 15 or (int(major) == 15 and int(minor) < 2):
-                raise OrchestratorError('cephadm only supports octopus (15.2.0) or later')
+            version_error = self._check_target_version(version)
+            if version_error:
+                raise OrchestratorError(version_error)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/upgrade.py:137]: why not let `_check_target_version` raise directly?

File: src/pybind/mgr/cephadm/upgrade.py [hunk 4]
```diff
@@ -293,7 +293,8 @@ def _do_upgrade(self):
                 })
                 return
             self.upgrade_state.target_id = target_id
-            self.upgrade_state.target_version = target_version
+            # extract the version portion of 'ceph version {version} ({sha1})'
+            self.upgrade_state.target_version = target_version.split(' ')[2]
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/upgrade.py:298]: this only works, if `target_version` is set by the newly upgraded MGR. If the `self.upgrade_state.target_version` was set by the old MGR, `self.upgrade_state.target_version` will be deserialized as the full string. 

I fear that we might need to cope with `self.upgrade_state.target_version` being set already .

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/cephadm/upgrade.py (hunk 1 above)
Comment (liewegas) [reply src/pybind/mgr/cephadm/upgrade.py:113]: I want to allow downgrade of point releases, though...

File: src/pybind/mgr/cephadm/upgrade.py (hunk 3 above)
Comment (liewegas) [reply src/pybind/mgr/cephadm/upgrade.py:137]: one caller raises (CLI error), the other caller calls _fail_upgrade()

File: src/pybind/mgr/cephadm/upgrade.py (hunk 4 above)
Comment (liewegas) [reply src/pybind/mgr/cephadm/upgrade.py:383]: added a compat kludge

**Comment by liewegas** [comment]
@sebastian-philipp Rebased on master... this is ready for a final review.

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
cc @yuriw

**Pull request merged**